### PR TITLE
Eng 19264 add allocator eecheck to simulate snapshot duplicate

### PR DESCRIPTION
**Pull request by kaikaizi**
- Added a multi-threaded EE test with well coordinated producer-consumer model to simulate possible race condition;
 - Fixed a multi-thread bug on chunk boundary under some (very) corner case
 - Added a write barrier (equivalent) to snapshot iterator, to read only when the whole allocator is done with **any** writing. Not sure if this would cure the failing system test; but since I don't have a perfect reproducer, this is the best I could come up with now.

### CODE REVIEW

File: tests/ee/storage/TableTupleAllocatorTest.cpp [hunk 1]
```diff
@@ -2151,6 +2153,66 @@ TEST_F(TableTupleAllocatorTest, TestFinalizer_Snapshot) {
...
+    // by deleting one tuple at a time, in the compacting direction.
+    // Synchronized perfectly with the snapshot iterator thread
+    // on each deletion
+    auto const deleting_thread = [&alloc, &addresses, &deleting_counter,
+         &iterating_counter, this] () {
+        int j = 0;
+        do {
+            for (int i = j + AllocsPerChunk - (j == 0 ? 2 : 1); i >= j; --i) {
+                alloc.remove_reserve(1);
+                alloc.template remove_add<truth>(const_cast<void*>(addresses[i]));
```
Comment (zeemanhe) [inline tests/ee/storage/TableTupleAllocatorTest.cpp]: add randomness the delete, such as multiple deletes?

File: tests/ee/storage/TableTupleAllocatorTest.cpp [hunk 2]
```diff
@@ -2151,6 +2153,66 @@ TEST_F(TableTupleAllocatorTest, TestFinalizer_Snapshot) {
     ASSERT_TRUE(verifier.ok(0));
 }
 
+TEST_F(TableTupleAllocatorTest, TestSimulateDuplicateSnapshotRead_mt) {
+    // test finalizer on iterator
+    using Alloc = HookedCompactingChunks<TxnPreHook<NonCompactingChunks<EagerNonCompactingChunk>, HistoryRetainTrait<gc_policy::always>>>;
+    using Gen = StringGen<TupleSize>;
+    Gen gen;
+    Alloc alloc(TupleSize);
+    array<void*, NumTuples> addresses;
```
Comment (zeemanhe) [inline tests/ee/storage/TableTupleAllocatorTest.cpp]: Have you tried to increase  NumTuples for the simulation?

File: tests/ee/storage/TableTupleAllocatorTest.cpp (hunk 1 above)
Comment (kaikaizi) [reply tests/ee/storage/TableTupleAllocatorTest.cpp]: Right now the deletion sequence is: each deletion will occur in the 2nd but last of 1st chunk, which triggers a compaction of the last tuple of 1st chunk to the deletion, conceptually like:
[0, 1, 2, ... N];
delete 1; -> [0, 2, ..., N];
delete 2; -> [0, 3, ..., N]
all the way to the tail, so that after the deletion sequence, only the last tuple remains, with value 0.
Since the snapshot is going on, in the txn allocator's memory, all the values in the allocator are 0.

The deletion thread and snapshot iterator are coordinated in a way that each deletion is followed by iterator advancement.

I have tried making the two thread un-coordinated, even by adding random sleep period. But that almost always works. The current sequence/coordination is the one that consistently revealed a multi-threaded bug.

Maybe there are other types of coordinated (not random, since random does not seem effective in revealing problems) would show other problems? I just don't know yet.

File: tests/ee/storage/TableTupleAllocatorTest.cpp (hunk 2 above)
Comment (kaikaizi) [reply tests/ee/storage/TableTupleAllocatorTest.cpp]: `NumTuples` is currently 8192. I can try increasing that; but with deterministic execution order that wouldn't reveal more problems.

File: tests/ee/storage/TableTupleAllocatorTest.cpp (hunk 1 above)
Comment (kaikaizi) [reply tests/ee/storage/TableTupleAllocatorTest.cpp]: Right now the deletion sequence is: each deletion will occur in the 2nd but last of 1st chunk, which triggers a compaction of the last tuple of 1st chunk to the deletion, conceptually like:
[0, 1, 2, ... N];
delete 1; -> [0, 2, ..., N];
delete 2; -> [0, 3, ..., N]
all the way to the tail, so that after the deletion sequence, only the last tuple remains, with value 0.
Since the snapshot is going on, in the txn allocator's memory, all the values in the allocator are 0.

The deletion thread and snapshot iterator are coordinated in a way that each deletion is followed by iterator advancement.

I have tried making the two thread un-coordinated, even by adding random sleep period. But that almost always works. The current sequence/coordination is the one that consistently revealed a multi-threaded bug.

Maybe there are other types of coordinated (not random, since random does not seem effective in revealing problems) would show other problems? I just don't know yet.
